src/rpc/client.cpp

```diff
@@ -248,11 +248,13 @@ UniValue RPCConvertValues(const std::string &strMethod, const std::vector<std::s
 UniValue RPCConvertNamedValues(const std::string &strMethod, const std::vector<std::string> &strParams)
 {
     UniValue params(UniValue::VOBJ);
+    UniValue positional_args(UniValue::VARR);
 
     for (const std::string &s: strParams) {
         size_t pos = s.find('=');
         if (pos == std::string::npos) {
-            throw(std::runtime_error("No '=' in named argument '"+s+"', this needs to be present for every argument (even if it is empty)"));
+            positional_args.push_back(rpcCvtTable.convert(strMethod, positional_args.size()) ? ParseNonRFCJSONValue(s) : s);
```

Comment:
I think we should probably still error if `params` has anything in it.

Either that, or if `params` happened to match the positional args, convert them to positional args.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/19762#discussion_r526505453

> I think we should probably still error if `params` has anything in it.
> 
> Either that, or if `params` happened to match the positional args, convert them to positional args.

It's already an error server side, so implementing this wouldn't have a practical effect and would complicate this PR. I can see benefits to doing fancier client side stuff, but I'd prefer to keep client changes minimal here, and add client features in a followup PR. If this relates to your other idea of making server validation less restrictive and merging parameters server-side, I also think that should be a followup PR.

In general, nothing in this PR gets in the way of providing merging features you are angling for, but this PR is intentionally only allowing basic (python-style) concatenation as a starting point

Reply:
For example, `bitcoin-cli getrawtransaction txid blockhash=blah 1` should probably fail, but unless I'm missing something, this will end up with the `1` as `verbosity` despite it being the 3rd on the command line option

Reply:
> For example, `bitcoin-cli getrawtransaction txid blockhash=blah 1` should probably fail, but unless I'm missing something, this will end up with the `1` as `verbosity` despite it being the 3rd on the command line option

I think you forgot `-named` in your example, but otherwise this is right and seems good to me. One of the nice things about named arguments is that that they bind unambiguously so you are free to place them anywhere. Only the order of positional arguments matters, so:

```
bitcoin-cli -named getrawtransaction txid 1 blockhash=blah
bitcoin-cli -named getrawtransaction txid blockhash=blah 1
bitcoin-cli -named getrawtransaction blockhash=blah txid 1
```

are all equivalent, and safe.

Meanwhile there is strict checking for validity of positional arguments. We always enforce that the first positional argument is the first RPC method argument, and the second positional argument is the second RPC method argument, etc. We explicitly disallow cases like the ones Jon cited https://github.com/bitcoin/bitcoin/pull/19762#pullrequestreview-478825831:

```
bitcoin-cli -named getrawtransaction txid=abc true
Parameter txid specified twice both as positional and named argument

bitcoin-cli -named getrawtransaction txid=abc 1
Parameter txid specified twice both as positional and named argument
```

where someone attempts to used named arguments to bump up positions of positional arguments. These could be allowed, but I think would be more dangerous and not actually useful.

Reply:
> re: [#19762 (comment)](https://github.com/bitcoin/bitcoin/pull/19762#discussion_r527223083)

To make this concrete, lets say you have an RPC with 4 parameters `a` `b` `c` `d` that you want to pass values `a` `b` `c` `d`.

This is allowed:

```sh
bitcoin-cli -named RPC a b c=c d=d
bitcoin-cli -named RPC c=c d=d a b
```

This is an error:

```sh
bitcoin-cli -named RPC a=a b=b c d
bitcoin-cli -named RPC a b=b c=c d
```

The client does not have parameter information so the errors are detected server side. This also means the client is less fragile and the server side error checking is compatible with other clients.

Reply:
You raise very solid points regarding JSON-RPC, python RPC and bitcoin-cli. Consider me fully convinced that further restrictions on positioning named before positional arguments should not be imposed. I was a bit too Python-biased in my assumptions. Thank you for the clear and comprehensive explanation.